cmd/thv-operator/pkg/vmcpconfig/converter.go

```diff
@@ -387,6 +397,166 @@ func (*Converter) convertCompositeTools(
+// merging them together and validating for duplicate names.
+func (c *Converter) convertAllCompositeTools(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+) ([]*vmcpconfig.CompositeToolConfig, error) {
+	ctxLogger := log.FromContext(ctx)
+
+	// Convert inline composite tools
+	inlineTools := c.convertCompositeTools(ctx, vmcp)
+
+	// Convert referenced composite tools if client is available
+	var referencedTools []*vmcpconfig.CompositeToolConfig
+	if c.k8sClient != nil && len(vmcp.Spec.CompositeToolRefs) > 0 {
+		var err error
+		referencedTools, err = c.convertReferencedCompositeTools(ctx, vmcp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert referenced composite tools: %w", err)
+		}
+	} else if len(vmcp.Spec.CompositeToolRefs) > 0 {
+		ctxLogger.Info("CompositeToolRefs specified but no Kubernetes client available, skipping referenced tools",
+			"vmcp", vmcp.Name,
+			"namespace", vmcp.Namespace,
+			"refCount", len(vmcp.Spec.CompositeToolRefs))
+	}
+
+	// Merge inline and referenced tools
+	allTools := make([]*vmcpconfig.CompositeToolConfig, 0, len(inlineTools)+len(referencedTools))
+	allTools = append(allTools, inlineTools...)
+	allTools = append(allTools, referencedTools...)
+
+	// Validate for duplicate names
+	if err := validateCompositeToolNames(allTools); err != nil {
+		return nil, fmt.Errorf("duplicate composite tool names: %w", err)
+	}
+
+	return allTools, nil
+}
+
+// convertReferencedCompositeTools fetches and converts referenced VirtualMCPCompositeToolDefinition resources.
+func (c *Converter) convertReferencedCompositeTools(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+) ([]*vmcpconfig.CompositeToolConfig, error) {
+	referencedTools := make([]*vmcpconfig.CompositeToolConfig, 0, len(vmcp.Spec.CompositeToolRefs))
+
+	for _, ref := range vmcp.Spec.CompositeToolRefs {
+		// Fetch the referenced VirtualMCPCompositeToolDefinition
+		compositeToolDef := &mcpv1alpha1.VirtualMCPCompositeToolDefinition{}
+		key := types.NamespacedName{
+			Name:      ref.Name,
+			Namespace: vmcp.Namespace,
+		}
+
+		if err := c.k8sClient.Get(ctx, key, compositeToolDef); err != nil {
+			if errors.IsNotFound(err) {
+				return nil, fmt.Errorf("referenced VirtualMCPCompositeToolDefinition %q not found in namespace %q",
+					ref.Name, vmcp.Namespace)
+			}
+			return nil, fmt.Errorf("failed to get VirtualMCPCompositeToolDefinition %q: %w", ref.Name, err)
+		}
+
+		// Convert the referenced definition to CompositeToolConfig
+		tool := c.convertCompositeToolDefinition(ctx, compositeToolDef)
+		referencedTools = append(referencedTools, tool)
+	}
+
+	return referencedTools, nil
+}
+
+// convertCompositeToolDefinition converts a VirtualMCPCompositeToolDefinition to CompositeToolConfig.
+func (*Converter) convertCompositeToolDefinition(
+	ctx context.Context,
+	def *mcpv1alpha1.VirtualMCPCompositeToolDefinition,
+) *vmcpconfig.CompositeToolConfig {
```

Comment:
the body of this function is almost the same as the body of the for loop in convertCompositeTools, can we reduce the code duplication?